File: pkg/dev_compiler/test/codegen_expected/language/try_catch4_test.js
dart_library.library('language/try_catch4_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__try_catch4_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const try_catch4_test = Object.create(null);
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  try_catch4_test.a = null;
  try_catch4_test.foo1 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.a = 8;
        return;
      } finally {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }
    }
  };
  dart.fn(try_catch4_test.foo1, VoidTodynamic());
  try_catch4_test.doThrow = function() {
    dart.throw(2);
  };
  dart.fn(try_catch4_test.doThrow, VoidTodynamic());
  try_catch4_test.foo2 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.a = 8;
        try_catch4_test.doThrow();
        return;
      } catch (e) {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }

    }
  };
  dart.fn(try_catch4_test.foo2, VoidTodynamic());
  try_catch4_test.foo3 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.doThrow();
      } catch (e) {
        try_catch4_test.a = 8;
        entered = true;
        return;
      }
 finally {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }
    }
  };
  dart.fn(try_catch4_test.foo3, VoidTodynamic());
  try_catch4_test.foo4 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.a = 8;
        break;
      } finally {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }
    }
  };
  dart.fn(try_catch4_test.foo4, VoidTodynamic());
  try_catch4_test.foo5 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.a = 8;
        try_catch4_test.doThrow();
        break;
      } catch (e) {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }

    }
  };
  dart.fn(try_catch4_test.foo5, VoidTodynamic());
  try_catch4_test.foo6 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.doThrow();
      } catch (e) {
        try_catch4_test.a = 8;
        entered = true;
        break;
      }
 finally {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }
    }
  };
  dart.fn(try_catch4_test.foo6, VoidTodynamic());
  try_catch4_test.foo7 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.a = 8;
        continue;
      } finally {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }
    }
  };
  dart.fn(try_catch4_test.foo7, VoidTodynamic());
  try_catch4_test.foo8 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.a = 8;
        try_catch4_test.doThrow();
        continue;
      } catch (e) {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }

    }
  };
  dart.fn(try_catch4_test.foo8, VoidTodynamic());
  try_catch4_test.foo9 = function() {
    let b = false;
    let entered = false;
    while (true) {
      if (entered) return b;
      b = dart.equals(8, try_catch4_test.a);
      try {
        try_catch4_test.doThrow();
      } catch (e) {
        try_catch4_test.a = 8;
        entered = true;
        continue;
      }
 finally {
        b = dart.equals(8, try_catch4_test.a);
        entered = true;
        continue;
      }
    }
  };
  dart.fn(try_catch4_test.foo9, VoidTodynamic());
  try_catch4_test.main_test = function() {
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo1());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo2());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo3());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo4());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo5());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo6());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo7());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo8());
    try_catch4_test.a = 0;
    expect$.Expect.isTrue(try_catch4_test.foo9());
  };
  dart.fn(try_catch4_test.main_test, VoidTodynamic());
  try_catch4_test.main = function() {
    for (let i = 0; i < 20; i++) {
      try_catch4_test.main_test();
    }
  };
  dart.fn(try_catch4_test.main, VoidTodynamic());
  // Exports:
  exports.try_catch4_test = try_catch4_test;
});

File: tests/language_2/try_catch4_test.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// Check that our SSA graph does have the try body a predecessor of a
// try/finally.
// VMOptions=--optimization-counter-threshold=10 --no-background-compilation

import "package:expect/expect.dart";

var a;

foo1() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      a = 8;
      return false;
    } finally {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

doThrow() {
  throw 2;
}

foo2() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      a = 8;
      doThrow();
      return false;
    } catch (e) {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

foo3() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      doThrow();
    } catch (e) {
      a = 8;
      entered = true;
      return false;
    } finally {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

foo4() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      a = 8;
      break;
    } finally {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

foo5() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      a = 8;
      doThrow();
      break;
    } catch (e) {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

foo6() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      doThrow();
    } catch (e) {
      a = 8;
      entered = true;
      break;
    } finally {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

foo7() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      a = 8;
      continue;
    } finally {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

foo8() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      a = 8;
      doThrow();
      continue;
    } catch (e) {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

foo9() {
  var b = false;
  var entered = false;
  while (true) {
    if (entered) return b;
    b = 8 == a; // This expression should not be GVN'ed.
    try {
      doThrow();
    } catch (e) {
      a = 8;
      entered = true;
      continue;
    } finally {
      b = 8 == a;
      entered = true;
      continue;
    }
  }
}

main_test() {
  a = 0;
  Expect.isTrue(foo1());
  a = 0;
  Expect.isTrue(foo2());
  a = 0;
  Expect.isTrue(foo3());
  a = 0;
  Expect.isTrue(foo4());
  a = 0;
  Expect.isTrue(foo5());
  a = 0;
  Expect.isTrue(foo6());
  a = 0;
  Expect.isTrue(foo7());
  a = 0;
  Expect.isTrue(foo8());
  a = 0;
  Expect.isTrue(foo9());
}

main() {
  for (var i = 0; i < 20; i++) {
    main_test();
  }
}
